From 8d583e71d7708b65b096661f6ab0dda71200f120 Mon Sep 17 00:00:00 2001
From: Michael Gill <gill@xilinx.com>
Date: Thu, 2 Jun 2016 18:21:19 -0700
Subject: [PATCH 742/827] staging: apf: Fix length for non-aligned(4K) DMA-BUF SG

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Iterate through DMA-BUF SG list and set length equal to
buffer size.

Signed-off-by: Michael Gill <gill@xilinx.com>
Signed-off-by: Christian Kohn <christian.kohn@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 963f3be71a875d95c12dfa0bf4ec50a5eff711aa)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/staging/apf/xilinx-dma-apf.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/staging/apf/xilinx-dma-apf.c b/drivers/staging/apf/xilinx-dma-apf.c
index 98dfd69..0ba536f 100644
--- a/drivers/staging/apf/xilinx-dma-apf.c
+++ b/drivers/staging/apf/xilinx-dma-apf.c
@@ -822,6 +822,10 @@ int xdma_submit(struct xdma_chan *chan,
 	dmahead->userflag = user_flags;
 	dmadir = chan->direction;
 	if (dp) {
+		struct scatterlist *sg;
+		int i;
+		unsigned int remaining_size = size;
+
 		if (!dp->is_mapped) {
 			dp->dbuf_attach = dma_buf_attach(dp->dbuf, chan->dev);
 			dp->dbuf_sg_table = dma_buf_map_attachment(
@@ -840,6 +844,15 @@ int xdma_submit(struct xdma_chan *chan,
 		sgcnt = dp->dbuf_sg_table->nents;
 		sgcnt_dma = dp->dbuf_sg_table->nents;
 
+		for_each_sg(sglist, sg, sgcnt, i) {
+			if (sg_dma_len(sg) > remaining_size) {
+				sg_dma_len(sg) = remaining_size;
+				remaining_size = 0;
+			} else {
+				remaining_size -= sg_dma_len(sg);
+			}
+		}
+
 		dmahead->userbuf = (void *)dp->dbuf_sg_table->sgl->dma_address;
 		dmahead->is_dmabuf = 1;
 	} else if (user_flags & CF_FLAG_PHYSICALLY_CONTIGUOUS) {
-- 
2.9.3

